docs/content/sips/0xx-spin-factors.md

```diff
@@ -0,0 +1,108 @@
+
+## Background
+
+Spin 1.0 shipped with a mechanism for extending the runtime environment with
+loosely-coupled host functionality called "host components". As Spin has
+evolved, more and more runtime functionality has fallen outside of the scope of
+this mechanism, leading to the code for these features spreading out over the
+Spin codebase. This not only makes it hard to read and understand these features
+but also raises the bar for unfamiliar developers trying to add new features.
+
+Separately, the introduction of the SpinKube project has made it clear that
+there will be (at least) two major embeddings of the Spin runtime environment:
+the Spin CLI (`spin up`) and SpinKube (`containerd-shim-spin`). In order to
+provide better integration into the Kubernetes ecosystem, some runtime feature
+implementations will diverge between these embeddings, making the loose coupling
+of features even more important.
+
+## Proposal
+
+The basic inversion of control approach used by `HostComponent`s will be
+redesigned and expanded into a new system called Spin Factors, where independent
+feature sets are organized into individual "factors". Some of the goals of this
+new system:
+
+- Expand the set integration points that a factor can use to hook into the
+  lifecycle of an application request. These will be driven by feature
+  requirements and may include e.g.:
+  - Runtime startup
+  - Application initialization
+  - Runtime config parsing
+  - Component pre-instantiation
+  - Component instantiation
+  - Cleanup (post-execution)
+
+- Allow loosely-coupled dependencies between "factors" (features). In order to
+  simplify implementation and reasoning about these dependencies the dependency
+  graph must be acyclic.
+
+## Implementation Plan
+
+The overall implementation plan will be to (in parallel):
+
+- Introduce a new `spin-factors` crate containing the basic framework.
+- Refactor existing Spin features (mostly existing `HostComponent`s) to use
+   this new framework.
+   - Framework features will be developed as needed to support this refactoring.
+- Refactor `spin-core`, `spin-trigger`, etc. to use `spin-factors`.
+   - Depending on subjective evaluation at this point, possibly merge
+   `spin-core` and `spin-factors`.
+
+## Implementation Details
+
+Based on initial prototyping, the following Rust types represent the starting
+point for `spin-factors` (note that some type details are elided for clarity):
+
+```rust
+pub trait Factor {
+    // Builder represents a type that can prepare request state for the factor.
+    type Builder: FactorBuilder<Self>;
+    // Data represents request state for the factor.
+    type Data;
+
+    // Init is the runtime startup lifecycle hook.
+    //
+    // The `InitContext` type here gives the factor the ability to update global
+    // engine configuration, most notably the `Linker`. This takes the place of
+    // `HostComponent::add_to_linker`.
+    fn init(ctx: InitContext) -> Result<()> {
```

Comment:
I would like to know more about `InitContext`. Could a rough definition of that be added to this code block?

Reply:
`InitContext` contains "whatever is needed to serve the requirements of this lifecycle hook". The impl PR, adapted from [the prototype](https://github.com/lann/spin-factors), has [more detail](https://github.com/fermyon/spin/pull/2519/files#diff-3896b08a07b6a026f5d2aca19afdee76e166070b8402329656dd5cb3726c05bfR28-R59).